feat: [M3-6646] -  Strip whitespace on blur for username and email textfields

mjac0bs opened the pull request:
## Description 📝
This change comes from a reporting of the "SOA email is not valid" validation error message on the Create Domains page (https://cloud.linode.com/domains/create) if a blank space is left in the email section after an inputted email.

The cause is that we don't currently strip extra whitespace on `TextFields`. For all email and username inputs, leading and trailing whitespace is stripped when a user focuses away from the input field.

## Major Changes 🔄
- Leading and trailing whitespace is stripped from user input in the following places:
   - Domains:
      - Create Domain `email` field (http://localhost:3000/domains/create)
      - Edit Domain drawer `email` field (http://http://localhost:3000/domains)
      - Edit SOA Record drawer `email` field (http://localhost:3000/domains/{id})
   - Account:
      - Add a User drawer `username` and `email` fields (http://localhost:3000/account/users)
      - User Profile `username` and `email` fields (http://localhost:3000/account/users/{username})
      - Display Settings `username` and `email` fields (http://localhost:3000/profile/display)
 - Managed: 
   - Add and Edit Credentials drawer `username` field for Managed accounts (http://localhost:3000/managed/credentials)
   - Add and Edit Contacts drawer `email` field for Managed accounts (http://localhost:3000/managed/contacts)
   - Note: we recently added trimming for SSH keys (M3-6473), and that functionality should still work in this PR. (http://localhost:3000/profile/keys)
- Adds a boolean`trimmed` prop to the `SingleTextFieldForm` and `TextField` components, which is used in the blur event handler to trim whitespace.


## Preview 📷

| Before  | After   |
| ------- | ------- |
| <video src="https://github.com/linode/manager/assets/114685994/59e19694-1547-4cbc-bce9-0e20a6aa4967"/> | <video src="https://github.com/linode/manager/assets/114685994/e4072a68-6aef-49a9-b0f2-cda25ee507bf"/> |

## How to test 🧪
1. **How to setup test environment?**
2. **How to reproduce the issue (if applicable)?**
   - Test any/all of the textfields mentioned above and note the error messages displayed when the user enters input with a leading or trailing spaces and then focuses away from the field.
4. **How to verify changes?**
   - Test the username, email, and SSH key fields throughout the app -- see above for a list. When entering input with leading or trailing whitespaces and focusing away from the field, no error message should display. When the form is submitted, the action should complete successfully.
   - Use the MSW for Managed.
5. **How to run unit or e2e tests?**
``` 
yarn test TextField
```

mjac0bs commented on packages/manager/src/features/Domains/CreateDomain/CreateDomain.tsx:
```diff
@@ -77,7 +78,7 @@ export const CreateDomain = () => {
     setSelectedDefaultNodeBalancer,
   ] = React.useState<NodeBalancer | undefined>(undefined);
 
-  const { values, ...formik } = useFormik({
```
Without un-destructuring this, L339 was throwing an error when I passed in `formik` as a param.

jaalah-akamai replied:
Another option is to leave this as is and in `handleTrimAndBlur` use a `Partial`:
```
formikProps: Partial<FormikProps<T>>
```
Then add some type guards before invoking the `setFieldValue` and `handleBlur` methods.

mjac0bs commented on packages/manager/src/features/Managed/Credentials/UpdateCredentialDrawer.tsx:
The diff in this file is primarily indentation; the code change is similar to `ContactsDrawer.tsx`, naming `formikProps` so that it can be used onBlur.

tyler-akamai commented:
> 
The Domain page is deleting the whitespace for me when I leave focus. However, the Account users page is not:
https://github.com/linode/manager/assets/139489745/6ba654e2-de1b-47f1-85f5-486e943721a9

tyler-akamai commented:
> > 
> 
> The Domain page is deleting the whitespace for me when I leave focus. However, the Account users page is not:
> 
>  Screen.Recording.2023-08-04.at.9.28.19.AM.mov

However the username trim is working on the "Add a User" drawer, so it seems like its just the email.

tyler-akamai commented:
> > > 
> > 
> > 
> > The Domain page is deleting the whitespace for me when I leave focus. However, the Account users page is not:
> > Screen.Recording.2023-08-04.at.9.28.19.AM.mov
> 
> However the username trim is working on the "Add a User" drawer, so it seems like its just the email.

If you console.log the email value within the render function, it is getting trimmed properly. It might be an issue with the value in the input field not getting updated properly.

mjac0bs commented:
Thanks @bnussman-akamai and @tyler-akamai. In both cases, it does look like the value itself is being trimmed, but the displayed text field value doesn't reflect that. This is odd because it's happening with two email fields, but not username fields, in places where both fields are in the same form, call essentially the same function onBlur, and then exhibit different behavior. I'm taking another 👀  to see what I missed here with emails.

mjac0bs commented on packages/manager/src/utilities/formikTrimUtil.tsx:
```diff
@@ -0,0 +1,14 @@
+import { FormikProps } from 'formik';
+
+export const handleTrimAndBlur = <T extends {}>(
```
I renamed to `handleFormikBlur` and added a JS doc. Do you have any thoughts on how to unit test the function? I attempted this in [be54e72](https://github.com/linode/manager/pull/9463/commits/be54e7272669a150abb840fe623b6a83e9bf8405) but couldn't figure out how to test without passing in the formikProps for the functions that `handleFormikBlur` calls, which needs to run to actually take care of the value update.

On a related note, I added `TextField.test.tsx` for unit test coverage of the `TextField` changes.

abailly-akamai replied:
i would mock those - something like this:

```
import { handleFormikBlur } from './formikTrimUtil';

describe('handleFormikBlur', () => {
  it('should trim leading and trailing space from a username TextField', () => {
    const setFieldValueMock = jest.fn();
    const handleBlurMock = jest.fn();
    const textfieldMock = {
      target: { name: 'firstNameField', type: 'text', value: '  John Doe  ' },
    } as React.FocusEvent<HTMLInputElement>;

    handleFormikBlur(textfieldMock, {
      handleBlur: handleBlurMock,
      setFieldValue: setFieldValueMock,
    });

    expect(setFieldValueMock).toHaveBeenCalledWith(
      'firstNameField',
      'John Doe'
    );
    expect(handleBlurMock).toHaveBeenCalledWith(textfieldMock);
  });

  it('should trim leading and trailing space from an email TextField', () => {
    const setFieldValueMock = jest.fn();
    const handleBlurMock = jest.fn();
    const emailFieldMock = {
      target: {
        name: 'emailField',
        type: 'email',
        value: '  johndoe@example.com  ',
      },
    } as React.FocusEvent<HTMLInputElement>;

    handleFormikBlur(emailFieldMock, {
      handleBlur: handleBlurMock,
      setFieldValue: setFieldValueMock,
    });

    expect(setFieldValueMock).toHaveBeenCalledWith(
      'emailField',
      'johndoe@example.com'
    );
    expect(handleBlurMock).toHaveBeenCalledWith(emailFieldMock);
  });
});

```

mjac0bs commented on packages/manager/src/features/Users/CreateUserDrawer.tsx:
```diff
@@ -58,15 +58,17 @@ class CreateUserDrawer extends React.Component<CombinedProps, State> {
           errorText={hasErrorFor('email')}
           label="Email"
-          onChange={this.onChangeEmail}
+          onBlur={this.handleTrimEmail}
```
A good question; when I looked back at this, I wasn't sure why I did that either. This was fixed in [f70d748](https://github.com/linode/manager/pull/9463/commits/f70d748b50d60a7ec7735a19434061400267900f).

mjac0bs commented on packages/manager/src/components/SingleTextFieldForm/SingleTextFieldForm.tsx:
```diff
@@ -40,6 +41,14 @@ export const SingleTextFieldForm = React.memo((props: Props) => {
   const [errors, setErrors] = React.useState<APIError[] | undefined>();
   const [success, setSuccess] = React.useState(false);
 
+  const handleTrim = (
```
Removed this function, as we can rely on `handleBlur` in `TextField` by passing down the `trimmed` prop, done in [f70d748](https://github.com/linode/manager/pull/9463/commits/f70d748b50d60a7ec7735a19434061400267900f).

mjac0bs commented:
> Thanks @bnussman-akamai and @tyler-akamai. In both cases, it does look like the value itself is being trimmed, but the displayed text field value doesn't reflect that. This is odd because it's happening with two email fields, but not username fields, in places where both fields are in the same form, call essentially the same function onBlur, and then exhibit different behavior. I'm taking another 👀 to see what I missed here with emails.

The cause of this was the `type=email` prop being set on some, but not all, email input fields. Browsers provide some input validation on fields with [a `type` of `email`](https://html.spec.whatwg.org/multipage/input.html#email-state-(type=email)), so the value was getting trimmed (even prior to this PR), but the display value was not updating because, as far as the component was concerned, there had been no changes to the value. The kind of hacky workaround for this is to update the value to a dummy value, then change it back to the input value to trigger the rerender. Not ideal and open to other thoughts on this, but I kept `type=email` for the sanitization and added [a check for the email type in the blur handler functions](https://github.com/linode/manager/pull/9463/commits/f70d748b50d60a7ec7735a19434061400267900f#diff-e300a581ed53f72758c3270dd71504c8db7c2f8bbf40636169b2a69d6d88a692R229-R232) to make sure the display value gets updated and the form behavior is consistent across fields.

tyler-akamai commented:
> 

Awesome, glad you were able to find it!

abailly-akamai reviewed (commented):
Looking great! actually wasn't able to find aforementioned inconsistent blur behavior and verified commented issues were now working. 

added a couple comments and a test suggestion. pretty close to being 💯 !

mjac0bs commented:
@abailly-akamai - Thanks for the feedback. A few changes since last viewed: 
- Cleaned up the blur event handlers
- Updated the `formilTrimUtil` unit test to use mock values (thank you, the `expect` statement was tripping me up there when I'd thought about how to mock functions)
- Searched through files again and discovered the Edit SOA Record drawer and Add Credentials drawer were missing trim behavior on blur and corrected that
- Partially reverted the change not to pass an onBlur handler (for username textfields, which are not validated with `type=email`) (https://github.com/mjac0bs/manager/commit/276cf5804d7bc0ac87a51654d56ae9da87304e49)... without the handler, form state doesn't update so although the trimming happens on the text field, the form can't submit without error. Let me know if you've got any concerns here or I've missed something.

In my testing, all listed form fields now have the same trimming behavior on blur and can submit a trimmed field without error.

bnussman-akamai reviewed (approved):
Trimming is working as expected where I tested!

carrillo-erik commented:
Looks like I'm reviewing this after all the minor tweaks have been addressed. All appropriate fields trim the whitespace as discussed in this ticket.
